pkg/store/detector.go

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package store
+
+import (
+	"strings"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+)
+
+// So far we found there is a bug in prometheus tsdb code when OOO is enabled.
+// This is a workaround to detect the issue when tsdb selects irrelevant matched data.
+// See https://github.com/thanos-io/thanos/issues/7481
+func detectCorruptLabels(lbls labels.Labels, matchers []storepb.LabelMatcher) bool {
+	for _, m := range matchers {
+		v := lbls.Get(m.Name)
+		if v == "" {
+			continue
+		}
```

Comment:
This branch means the time series doesn't have this label, but the matches (filters) have. The workflow should go to do a match against the default empty string. Otherwise, it would cause the following false negative.
```
lbls = {"a": "b"}

matches = [ {Name: "b", Type: LabelMatcher_EQ, Value: "x" }]
```

Reply:
emm, i think i added this due to a unit test failure, let me figure out what happened.

Reply:
test it here: https://github.com/databricks/thanos/pull/57